torch/sparse/semi_structured/semi_structured_sparse_tensor.py

```diff
@@ -0,0 +1,156 @@
+import torch
+
+
+class SemiStructuredSparseTensor(torch.Tensor):
+    """
+    This class implementes semi-structured (2:4) sparsity as a Tensor subclass.
+
+    Semi-structured sparsity describes a sparsity pattern where n in every 2n elements are sparse,
+    depending on the datatype. It is most commonly referred to as 2:4 sparsity or fine-grained
+    structured sparsity.
+
+    compressed_tensor is a tensor that stores both the kept elemenets and the metadata mask
+    These two are stored next to each other in one contiguous tensor.
+
+    compressed tensor = [ kept elemetns of original tensor |   mask_metadata     ]
+
+    For an original tensor of size (m, k) we expect the first m * k // 2 elements to be the kept elements
+    The rest of the tensor is metadata.
+
+    This tensor subclass also has a cslt object, which is set if pytorch is built with cuSPARSELt support.
+    In this case, we store some additional metadata containg the sparse matrix descriptor in order for
+    faster matmul performance.
+
+    When this cslt object is not set, we will default to using CUTLASS kernels and _structured_sparse_linear.
+    """
+
+    # When fuse_transpose is set to True, we fuse a .T into the cuSPASRELt matmul operation.
+    # We do this in a bit of a hacky manner, by just creating a transposed matrix and then setting
+    # the output order to be Column major instead of Row major.
+    fuse_transpose = True
+
+    @staticmethod
+    def __new__(cls, custom_shape, compressed_tensor, cslt, transposed):
```

Comment:
Let's throw a [prototype](https://pytorch.org/docs/stable/index.html) warning in here (similar to [`MaskedTensor`](https://github.com/pytorch/pytorch/blob/main/torch/masked/maskedtensor/core.py#L156-L159)). This is a normal lifecycle for new features and we communicate to users that we're iteratively building this out.